src/Functions/seriesDecomposeAnomaliesDetection.cpp

```diff
@@ -0,0 +1,277 @@
+                throw Exception(ErrorCodes::BAD_ARGUMENTS, "The fourth argument of function can only be 'tukey' or 'ctukey'.");
+
+            if (kind == "tukey")
+            {
+                min_percentile = 25;
+                max_percentile = 75;
+            }
+        }
+
+        ColumnsWithTypeAndName decompose_cols{arguments[0]};
+
+        if (seasonality == -1)
+        {
+            ColumnsWithTypeAndName columns{arguments[0]};
+            auto get_seasonality = FunctionFactory::instance().get("seriesPeriodDetectFFT", context)->build(columns);
+            auto get_seasonality_res = get_seasonality->execute(columns, std::make_shared<DataTypeUInt64>(), input_rows_count);
+
+            decompose_cols.push_back({get_seasonality_res, std::make_shared<DataTypeUInt64>(), "period"});
+        }
+        else
+        {
+            auto seasonality_col = arguments[2].column->convertToFullColumnIfConst();
+            decompose_cols.push_back({seasonality_col, arguments[2].type, "period"});
+        }
+
+        auto get_decompose = FunctionFactory::instance().get("seriesDecomposeSTL", context)->build(decompose_cols);
+        auto decompose_res_type = std::make_shared<DataTypeArray>(std::make_shared<DataTypeArray>(std::make_shared<DataTypeFloat32>()));
+        auto get_decompose_res = get_decompose->execute(decompose_cols, decompose_res_type, input_rows_count);
+
+        const ColumnArray * decompose_res_array = checkAndGetColumn<ColumnArray>(get_decompose_res.get());
+        auto root_decompose_data = &decompose_res_array->getData();
+        const IColumn::Offsets & root_decompose_offsets = decompose_res_array->getOffsets();
+
+        const ColumnArray * decompose_data_array = checkAndGetColumn<ColumnArray>(root_decompose_data);
+        const IColumn::Offsets & nested_decompose_offsets = decompose_data_array->getOffsets();
+        const IColumn & src_data = decompose_data_array->getData();
+        const ColumnVector<Float32> * src_data_concrete = checkAndGetColumn<ColumnVector<Float32>>(src_data);
+        const PaddedPODArray<Float32> & decompose_data = src_data_concrete->getData();
+
+        ColumnArray::Offset prev_decompose_offset = 0;
+
+        auto res = ColumnFloat64::create();
+        auto & res_data = res->getData();
+
+        ColumnArray::ColumnOffsets::MutablePtr res_col_offsets = ColumnArray::ColumnOffsets::create();
+        auto & res_col_offsets_data = res_col_offsets->getData();
+
+        auto root_offsets = ColumnArray::ColumnOffsets::create();
+        auto & root_offsets_data = root_offsets->getData();
+
+        for (size_t i = 0; i < root_decompose_offsets.size(); i++)
+        {
+            //extract residual component of the decomposed series
+            ColumnArray::Offset residual_start_offset = nested_decompose_offsets[prev_decompose_offset + 1];
+            ColumnArray::Offset residual_end_offset = nested_decompose_offsets[prev_decompose_offset + 2];
```

Comment:
AFAIK from [here](https://clickhouse.com/docs/en/sql-reference/functions/time-series-functions#seriesdecomposestl), as a result from seriesDecomposeSTL we have an array of arrays, residual component is the third array, but as I see here, we take the second array? (I may be wrong)